pallets/data-preservers/src/lib.rs

```diff
@@ -550,22 +613,132 @@ pub mod pallet {
+                assigner_param,
+            )?;
+
+            profile.assignment = Some((para_id, witness));
+            Profiles::<T>::insert(profile_id, profile);
+
+            Self::deposit_event(Event::AssignmentStarted {
+                profile_id,
+                para_id,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(8)]
+        #[pallet::weight(0)]
+        pub fn stop_assignment(
+            origin: OriginFor<T>,
+            profile_id: T::ProfileId,
+            para_id: ParaId,
+        ) -> DispatchResultWithPostInfo {
+            let assigner = T::AssignmentOrigin::ensure_origin(origin, &para_id)?;
+
+            let mut profile = Profiles::<T>::get(profile_id).ok_or(Error::<T>::UnknownProfileId)?;
+
+            let Some((assignment_para_id, assignment_witness)) = profile.assignment.take() else {
+                Err(Error::<T>::ProfileNotAssigned)?
+            };
+
+            if assignment_para_id != para_id {
+                Err(Error::<T>::WrongParaId)?
+            }
+
+            T::AssignmentPayment::try_stop_assignment(
+                assigner,
+                profile.account.clone(),
+                assignment_witness,
+            )?;
+
+            Profiles::<T>::insert(profile_id, profile);
+
+            {
+                let mut assignments = Assignments::<T>::get(para_id);
+                let Ok(position) = assignments.binary_search(&profile_id) else {
+                    Err(Error::<T>::ProfileNotAssigned)?
+                };
+
+                assignments.remove(position);
+
+                Assignments::<T>::insert(para_id, assignments);
+            }
+
+            Self::deposit_event(Event::AssignmentStopped {
+                profile_id,
+                para_id,
+            });
+
+            Ok(().into())
+        }
     }
 
     impl<T: Config> Pallet<T> {
+        pub fn assignments_profiles(para_id: ParaId) -> impl Iterator<Item = Profile<T>> {
+            Assignments::<T>::get(para_id)
+                .into_iter()
+                .filter_map(Profiles::<T>::get)
+                .map(|profile| profile.profile)
+        }
+
         /// Function that will be called when a container chain is deregistered. Cleans up all the storage related to this para_id.
         /// Cannot fail.
         pub fn para_deregistered(para_id: ParaId) {
-            BootNodes::<T>::remove(para_id);
+            Assignments::<T>::remove(para_id);
         }
 
         pub fn check_valid_for_collating(para_id: ParaId) -> DispatchResult {
-            // To be able to call mark_valid_for_collating, a container chain must have bootnodes
-            if Pallet::<T>::boot_nodes(para_id).len() > 0 {
-                Ok(())
-            } else {
-                Err(Error::<T>::NoBootNodes.into())
+            if Self::assignments_profiles(para_id)
+                .filter(|profile| profile.mode == ProfileMode::Bootnode)
+                .count()
+                == 0
```

Comment:
To stop iterating after first match:

```suggestion
                .any(|profile| profile.mode == ProfileMode::Bootnode)
```